### PR TITLE
[refactor](Nereids) New expression extractor for partitions pruning

### PR DESCRIPTION
**Pull request by XuPengfei-1020**
## Proposed changes

An exception throw in TryEliminateUninterestedPredicates, for this case
```sql
 CREATE TABLE `tbltest` (
  `id` INT NULL,
  `col2` VARCHAR(255) NULL,
  `col3` VARCHAR(255) NULL,
  `dt` DATE NULL
) ENGINE=OLAP
DUPLICATE KEY(`id`, `col2`)
PARTITION BY RANGE(`dt`)
(PARTITION p20240617 VALUES [('2024-06-17'), ('2024-06-18')))
DISTRIBUTED BY HASH(`id`) BUCKETS 10
PROPERTIES (
"replication_allocation" = "tag.location.default: 1"
);

select * from tbltest
where
  case
    when col2 = 'xxx' and col3='yyy' then false -- note this is not about partition column
    when col2 in ('xxx') then false
    when col2 like 'xxx%' then false
    else true
  end
```

The `CaseWhen` require children should be `WhenClause`, TryEliminateUninterestedPredicates maybe rewrite the WhenClause to true/false predicate, and cause this exception:
```
ERROR 1105 (HY000): errCode = 2, detailMessage = The children format needs to be [WhenClause+, DefaultValue?]
```

Original extractor(TryEliminateUninterestedPredicates.java) caused some errors while try to derive the expressions which can be used for pruning partitions.
I tried to write a new extractor(and with unit tests) for pruning partitions, it is more simple and reliable (I think).

The theory of extractor is pretty simple:
A:Sort the expression in two kinds:
  1. evaluable-expression (let's mark it as E).
    Expressions that can be evaluated in the partition pruning stage.
    In the other word: not contains non-partition slots or deterministic expression.
  2. un-evaluable-expression (let's mark it as UE).
    Expressions that can NOT be evaluated in the partition pruning stage.
    In the other word: contains non-partition slots or deterministic expression.

B: Travel the predicate, only point on AND and OR operator, following the rule:
  (E and UE) -> (E and TRUE) -> E
  (UE and UE) -> TRUE
  (E and E) -> (E and E)
  (E or UE) -> TRUE
  (UE or UE) -> TRUE
  (E or E) -> (E or E)

### CODE REVIEW

**Comment by 924060929** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/PartitionPruneExpressionExtractor.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+            }
+
+            // todo: Should we consider un-deterministic functions as un-evaluable expression?
+            // if (originExpr instanceof Function) {
+            //     Function function = (Function) originExpr;
+            //     if (Env.getCurrentEnv().isNondeterministicFunction(function.getName())) {
+            //         parentContext.containsUnEvaluableExpression = true;
+            //         return originExpr;
+            //     }
+            // }
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/PartitionPruneExpressionExtractor.java:189]: not need, some Nondeterministic functions are foldable and others are not. FoldConstantRuleOnFE can detect it.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/PartitionPruneExpressionExtractor.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        this.expressionEvaluableDetector = new ExpressionEvaluableDetector(interestedSlots);
+    }
+
+    /**
+     * Extract partition prune expression from predicate
+     */
+    public static Expression extract(Expression predicate,
+                                     Set<Slot> partitionSlots,
+                                     CascadesContext cascadesContext) {
+        predicate = predicate.accept(FoldConstantRuleOnFE.VISITOR_INSTANCE, null);
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/PartitionPruneExpressionExtractor.java:77]: new an ExpressionRewriteContext as context parameter

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/PartitionPruneExpressionExtractor.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    /**
+     * The detector only indicate that whether a predicate contains interested slots or not,
+     * and do not change the predicate.
+     */
+    @VisibleForTesting
+    public static class ExpressionEvaluableDetector extends DefaultExpressionRewriter<Context> {
+        private final Set<Slot> partitionSlots;
+
+        public ExpressionEvaluableDetector(Set<Slot> partitionSlots) {
+            this.partitionSlots = Objects.requireNonNull(partitionSlots);
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/PartitionPruneExpressionExtractor.java:163]: and a simple exception message here, e.g.
```java
this.partitionSlots = Objects.requireNonNull(partitionSlots, "partitionSlots can not be null);
```

**Comment by XuPengfei-1020** [comment]
run buildall

**Review by 924060929 — APPROVED** [review]

**Comment by 924060929** [comment]
run buildall

**Comment by morrySnow** [comment]
could u add a case that can reproduce this problem?

**Comment by XuPengfei-1020** [comment]
> could u add a case that can reproduce this problem?

The case that can reporduce the issue at above, "Proposed changes" chapter (Added by @924060929 ).

**Review by morrySnow — APPROVED** [review]

**Pull request merged**